### PR TITLE
add python script to generate config based on gitlab group

### PR DESCRIPTION
**Pull request by aschleifer**
@tony this is the python version of the script we added with #336 for #334 .

It has the same usages as for the bash script, but it also validates that it can write the file and will ask the user if it should overwrite an existing config.

As I said before I'm not a python expert, so if you have any improvement ideas please let me know.

Also I wasn't sure about any style guides this project follows. If you name me the guide I'm happy to format the code accordingly.

Best regards
Segaja

### CODE REVIEW

**Comment by aschleifer** [comment]
Ok, I saw that github automatically ran flake8 against the changes and I fixed the issues it showed.

**Comment by aschleifer** [comment]
I must say I'm not a fan of all changes the `[pre-commit.ci]` is doing ( https://github.com/vcs-python/vcspull/pull/337/commits/ae27f03a0579529d124bef3db08467393aab16b9 ). Some of the formattings just make the code harder to read now, but ok.

**Comment by tony** [comment]
Yeah, it formats to to what `black` formatter's settings are

this is becoming more and more common in open source projects (not the preformat, but running `black` on the code)

I will look at this tonight / tomorrow

**Comment by aschleifer** [comment]
> Yeah, it formats to to what `black` formatter's settings are
> 
> this is becoming more and more common in open source projects (not the precommit, but running `black` on the code).

I'm ok with running a formatter in a project to agree on the formatting. I'm just saying I myself are not too happy with one of the formatting options, but I can live with it. 


> you could also rebase it into your changes and squash as you'd like in the mean time

Nah, I'm fine with leaving it as it is right now. Unless you want me to rebase this all into one perfect commit which brings in the script in one clean commit.

**Comment by tony** [comment]
> Unless you want me to rebase this all into one perfect commit which brings in the script in one clean commit.

Based on the commits in the PR, i will likely do "squash and merge" which will have the same effect when merged to the trunk

**Comment by tony** [comment]
@aschleifer Is this good to merge for now? Anything else you'd like me to add?

**Comment by aschleifer** [comment]
> @aschleifer Is this good to merge for now? Anything else you'd like me to add?

From my side this is good to go. I guess in the end the only thing missing is the already mentioned example entry.

**Comment by tony** [comment]
@aschleifer This sounds good.  Thank you for this!

Merging this in

This coming week I will make a release that ships a root command with `vcspull sync` and through the next week(s), publish fixes for the bugs.

File: scripts/generate_gitlab.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    help="path to the target config file (default: ./vcspull.yaml)",
+    dest="config_file_name",
+    required=False,
+    default="./vcspull.yaml",
+)
+
+args = vars(parser.parse_args())
+gitlab_host = args["gitlab_host"]
+gitlab_namespace = args["gitlab_namespace"]
+config_filename = args["config_file_name"]
```
Comment (dvzrv) [inline scripts/generate_gitlab.py:38]: You might just want to create a [pathlib.Path](https://docs.python.org/3.10/library/pathlib.html) here, as it has easy object based handling (e.g. [Path.is_file()](https://docs.python.org/3.10/library/pathlib.html#pathlib.Path.is_file))

File: scripts/generate_gitlab.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+try:
+    if os.path.isfile(config_filename):
+        result = input(
+            "The target config file (%s) already exists, \
```
Comment (dvzrv) [inline scripts/generate_gitlab.py:43]: As this codebase is for python >= 3.7, you can just use f-strings [PEP0498](https://www.python.org/dev/peps/pep-0498/). E.g. `f"foo {my_variable}"`

File: scripts/generate_gitlab.py (hunk 1 above)
Comment (tony) [reply scripts/generate_gitlab.py:38]: @dvzrv @aschleifer  would either of you like to make a follow up PR improving upon this?

**Comment by tony** [comment]
@aschleifer Added to https://vcspull.git-pull.com/config-generation.html#python-version

Similar to #336, I included my demo usage there.

**Comment by tony** [comment]
@aschleifer P.S. Is it your desire to have the scripts included in the python package itself?  Either inside the directory or available in the `bin/` packages?

**Comment by aschleifer** [comment]
> @aschleifer P.S. Is it your desire to have the scripts included in the python package itself? Either inside the directory or available in the `bin/` packages?

I think at the end I will try to add this as a "native command" in vcspull and then we can remove these scripts.

File: scripts/generate_gitlab.py (hunk 1 above)
Comment (aschleifer) [reply scripts/generate_gitlab.py:38]: @tony I will do a follow up PR soon to address this point.

File: scripts/generate_gitlab.py (hunk 2 above)
Comment (aschleifer) [reply scripts/generate_gitlab.py:43]: @tony how do you stand on the topic of injecting variables into strings? is there one formatting / style you would like to see used in the python code?

**Pull request merged**